### DISCUSSION TITLE
Notifications are here

### DISCUSSION THREAD

**Discussion by jocmp** [jocmp/capyreader#424] [Announcements]
The first iteration of notifications is complete and available in the latest build, [2024.10.1060-dev](https://github.com/jocmp/capyreader/releases/tag/2024.10.1060-dev). I’ll push these changes to the Google Play Store later this week as well. I want to give a brief overview of how notifications work in Capy since behaviors compared to other apps like Feeder or ReadYou and user expectations may differ.

Notifications are configurable from Settings under Settings → General → Refresh. The refresh setting controls how frequently feeds are fetched and in turn control how often notifications appear. 

The notification item is clickable after accepting the notification permission. From here you can select as many or as few feeds as possible.

<img width="400px" src="https://github.com/user-attachments/assets/e6c7dced-68fd-44ad-a616-79c8e8c5fdab" alt="Notification settings" />

Notifications will only count new, unread articles that have arrived since the refresh. Each notification is sent at a feed level, so a notification might say “1 new articles from 9to5Google” if 1 new article has arrived since the last refresh.

On click, the notification will open to the unread filter for the feed. From there you can tap into the article or articles that have arrived.

<img width="400px" src="https://github.com/user-attachments/assets/3a58dbbf-cae7-422d-82ef-3246b0ff9721" alt="Capy Reader notification" />

## Battery Optimization
Please note that refresh frequency is still relative based on your device’s vendor and how they handle background refresh. Pixel devices for instance may choose to skip a 15 minute refresh if the app hasn’t been used in a while to conserve battery life. For Android 6+ can be [overridden in app settings](https://dontkillmyapp.com/general#user-solution) but is not necessary in my testing. 

Visit the [Don’t kill my app!](https://dontkillmyapp.com/) site to see how your device vendor stacks up and how they handle background refreshes. 

## Future Changes

At some point I’d like to add more ways to access notifications from a feed level with a bell icon. For now I want to keep the changes smallish and get your feedback as soon as possible. Try it out, let me know what you think!

**Comment by PhilC813** [comment]
With the new version "2024.10.1060-dev", I see the new "Refresh Feeds" menu under "General" but none for notifications specifically. I have the notification permission given for the app at the Android level.

**Reply by jocmp** [reply]
Can you post a screenshot of what you're seeing? The notifications section should show up under General if you have the permission enabled so I'm not sure what you might be running into.

**Reply by PhilC813** [reply]
Here is the "General" section in the app:
![Screenshot_20241014_223744](https://github.com/user-attachments/assets/7e09f858-6f8e-4c68-9a8c-3abfe3b472b1)

The permissions provided to the app:
![Screenshot_20241014_223938](https://github.com/user-attachments/assets/9450ec60-573f-4ca2-8645-44f8ce8abab5)

I get updates with Obtainium linked to the GitHub project.
![Screenshot_20241014_223748](https://github.com/user-attachments/assets/e68886f2-7965-4a4b-9652-8b105ea04731)

**Reply by PhilC813** [reply]
(I can always try a reinstall if you'd like, but I would tend to think that it has something to do with the code)

**Reply by jocmp** [reply]
Thanks @Philippe-Choquette. I agree that these all point to a code issue though I'm noticing some old spacing above the "Clear all articles" button that was also changed in the latest version. Can you do one more thing and check the version that the app says under Settings > About? Here's what I'm seeing in comparison

| General | About |
|--|--|
| <img src="https://github.com/user-attachments/assets/f3db8e94-709a-4a3e-a7db-292a408a33ce" width="400px"> |  <img src="https://github.com/user-attachments/assets/6c3cd936-bdf6-447e-ba20-fc473170c969" width="400px"> |

**Reply by PhilC813** [reply]
I'm decently proficient with tech but sometimes you overlook simple things... Thanks for pointing this out. The app was indeed showing v1059, and so was the "App info" screen in Android settings. I reinstalled the update with Obtainium and I now see the notifications menu in Capy. I'll blame Obtainium for this one 😅

**Reply by jocmp** [reply]
Yep seems like some wires got crossed with the packaging. I'm glad it finally installed correctly!

**Comment by PhilC813** [comment]
Based on my first try with one of my feeds, notifications work very well. I like that the layout is kept simple and that tapping them opens the unread filter for the relevant feed; nice touch.

I would expect that some people will want notifications for individual articles however, similar to how it's implemented in services like Inoreader. If you ever venture this far, I believe Android allows notifications to be split into groups, which would allow you to keep the feeds separate; this would be an improvement over the implementation of some other RSS readers.

**Reply by jocmp** [reply]
thanks for your input here. I took another pass at the grouping. I updated notifications to support article nav instead of feeds in the latest build. That will go out shortly. Here's what it looks like
 
<img src="https://github.com/user-attachments/assets/015d6009-7649-45fe-ac35-ae8a788bd746" width="400px">

**Reply by PhilC813** [reply]
"2 steps ahead"
That looks absolutely amazing

**Reply by PhilC813** [reply]
The title of articles is truncated in the notifications and it makes it difficult to get a sense of what articles are actually about. Would there be a way to improve this?

| Collapsed | Expanded |
| --- | --- |
| ![Screenshot_20241017_120411](https://github.com/user-attachments/assets/7e0b57f1-f7a5-4491-a2f8-e0b0857f3e97) | ![Screenshot_20241017_120321](https://github.com/user-attachments/assets/da7f5ee9-2b31-4abe-b09c-abc83e0e4ff2) |

**Reply by jocmp** [reply]
I can do some digging. Nothing comes to mind short of swapping the order around. Do you have any apps in mind that do notifications well? I can check those out as reference. I think Gmail has a good pattern for dealing with information density.

**Reply by PhilC813** [reply]
I haven't used the Gmail app for some time but being a Google app, it must have a decent notifications layout.

I installed a few RSS apps for comparison, but from those that didn't need to create an account, none had what I would consider to be a good layout; RSS apps are still a rather small market.

So I've tested a few mainstream news apps instead:
| App | Collapsed | Expanded |
| --- | --- | --- |
| WSJ | ![WSJ - Collapsed notification](https://github.com/user-attachments/assets/c5ceed01-06a3-4f7b-ab3b-a14ccb2df6f7) | ![WSJ - Expanded notification](https://github.com/user-attachments/assets/09157b01-2f99-4246-a0ab-508becf81ccb) |
| CNN | ![CNN - Collapsed notification](https://github.com/user-attachments/assets/3dcb34bf-f7f6-41e6-baa4-0dea015bbfd5) | ![CNN - Expanded notification](https://github.com/user-attachments/assets/036848ba-ec1f-4d18-9e25-b13e9dd4f52a) |
| Opera (web browser) | ![Opera News - Collapsed notification](https://github.com/user-attachments/assets/790f22ef-2b26-4ccd-9547-2bb42e247b65) | ![Opera News - Expanded notification](https://github.com/user-attachments/assets/74c6acfe-143e-484b-914a-73fee4a52940) |

I would say that the WSJ app has the best notification layout. They seem to be using the notification content field instead of the title one so as to include the full article title. CNN's layout is similar, but they make use of the notification title field to add the type of news (which isn't relevant in Capy's case since there are already notification groups to differentiate the feeds).

**Reply by jocmp** [reply]
hey thanks for the screenshots, this is super helpful. I think I found a way to format long titles like the WSJ via the [BigTextStyle](https://developer.android.com/reference/android/app/Notification.BigTextStyle) option. Here's a comparison.

| Before (Collapsed) | After (Collapsed) | After (Expanded) |
|:--|:--|:--|
| ![before_expanded](https://github.com/user-attachments/assets/49ac4371-cbe3-45ea-93d2-5fda69120533) | ![after_collapsed](https://github.com/user-attachments/assets/40679daf-cb49-42be-adc2-3c001fc0c804) | ![after](https://github.com/user-attachments/assets/a845ef69-1a5d-4c05-a1f9-43e91b80beff) |

**Reply by PhilC813** [reply]
The BigTextStyle does seem like the right option. The padding looks larger on your screenshots that may just be the difference between our Android manufacturer builds. I'll let you know on the next release! Thank you so much!

**Reply by PhilC813** [reply]
The new notification layout is perfect! It allows you to more easily evaluate articles without having to open the app.

If I could suggest one last thing:
When you open an article from the notification, and tap the back button to access the article list, it shows the feed with the "All" filter when it should ideally be the "Unread" filter instead.

**Reply by jocmp** [reply]
Sounds good to me. I'll tack this on to the next build.

**Reply by PhilC813** [reply]
A "Mark as read" button for the notification could be useful too when you know that the article doesn't interest you. Otherwise, when you open the app, you'll see many unread articles on the relevant feed, prompting you to check, only to notice that these are all articles you purposely ignored.

That makes every notification larger when expanded tho, so less "minimalist" if it matters to you.

Next to "Mark as read", you could also add a "Star" button; I think those would be the main useful options. The CNN app notifications have a "Share" button but I think most people would at least open the article before considering sharing with someone, and their "View" button is just redundant.

**Reply by jocmp** [reply]
I've been thinking the same for a "Mark as read" button at least. Tracking via https://github.com/jocmp/capyreader/issues/462

**Reply by PhilC813** [reply]
> When you open an article from the notification, and tap the back button to access the article list, it shows the feed with the "All" filter when it should ideally be the "Unread" filter instead.

I'm pretty sure that this was implemented but since the rework of the sidebar with the move from the dropdown to the icons, it now goes back to the "All" view.

**Reply by jocmp** [reply]
Oof yep, this is a regression. Initially I switched it back to "All" after seeing unexpected behavior with horizontal scrolling. Basically the article would disappear immediately after it was selected. I believe I fixed that behavior with the latest update, so I can re-add the "unread" status selection.

Tracking here

- https://github.com/jocmp/capyreader/issues/846

**Reply by PhilC813** [reply]
> Basically the article would disappear immediately after it was selected.

Oh I understand. Thank you checking this out!
